### PR TITLE
Fix data handling during merging and unmerging of CSM instances

### LINKED ISSUE #491: CSM merge does not consider assigned data
As discussed with @marscher last Friday, I tested if merging two CSMs transfers data to the child CSM to the parent CSM:

~~~ python
from weldx import CoordinateSystemManager

csm_1 = CoordinateSystemManager("r")
csm_1.create_cs("a", "r", coordinates=[1, 3, 4])
csm_1.create_cs("b", "r", coordinates=[5, 5, 5])

csm_2 = CoordinateSystemManager("x")
csm_2.create_cs("a", "x", coordinates=[0, 3, 0])
csm_2.create_cs("y", "x", coordinates=[5, 0, 0])

csm_2.assign_data([[1, 2, 3], [4, 5, 6]], "data", "y")

print(csm_2.data_names)

csm_1.merge(csm_2)

print(csm_1.data_names)

unmerged = csm_1.unmerge()
print(unmerged[0].data_names)
~~~

output:

~~~
['data']
[]
[]
~~~

As you can see, this is not the case (unmerging fails too). As far as I remember, this should be a minor implementation issue. The data itself should be copied, but it isn't registered in the internal structure. I will check if unmerging is also affected.

### PR DESCRIPTION
**Pull request by vhirtham**
## Changes

- CSM `.merge` and `.unmerge` now correctly merges and restores data attached to a CS. 
- remove internal variable `._data`

## Related Issues

Closes #491

## Checks

- [x] updated CHANGELOG.md
- [x] updated tests
- [x] updated doc/
- [x] ~~update example/tutorial notebooks~~

### CODE REVIEW

**Comment by vhirtham** [comment]
This was a bit tricky due to some annoying hidden side-effects but now it is working to some degree. I am not sure if the serialization works for all cases since not everything is tested but we will touch this when working on the graph serialization anyways. I guess this will cover everything @marscher needs to work on the dashboards.

I will add some edge case tests and clean everything up. Afterwards, this is ready.

File: weldx/transformations/cs_manager.py [hunk 1]
```diff
@@ -43,14 +43,55 @@ class CoordinateSystemManager:
...
+            if self.nodes is None:
+                if other.nodes is not None:
+                    return False
+            else:
+                if other.nodes is None:
+                    return False
+                for node in self.nodes:
+                    if node not in other.nodes:
+                        return False
+            return True
```
Comment (vhirtham) [inline weldx/transformations/cs_manager.py:92]: This can be simplified using `compare_nested` as follows:

~~~ python
        def __eq__(self, other):
            if not isinstance(other, CoordinateSystemManager.SubsystemData):
                raise NotImplementedError
            return util.compare_nested(self.__dict__, other.__dict__)
~~~

Problem is that compare nested requires that the order of list items is identical, which is not guaranteed during deserialization. I tried using sets instead, but it seems that `compare_nested` does not support them yet (see #495 )

File: weldx/util/util.py [hunk 2]
```diff
@@ -294,6 +291,18 @@ def compare_nested(
...
+            if isinstance(v, set):
+                return [], ((i, j) for i, j in enumerate(sorted(list(v))))
+            return iterutils.default_enter(p, k, v)
+
+        try:
+            a = iterutils.remap(a, enter=_enter_replace_sets)
+            b = iterutils.remap(b, enter=_enter_replace_sets)
+        except TypeError:
+            raise TypeError("either a or b are not a nested data structure.")
+
```
[the remark this answers is not in the document]
Comment (vhirtham) [reply weldx/util/util.py:305]: Is principle, lists and sets should always compare false against each other, so I am fine with that. However, I am not sure if this version is behaving as it should. (See my comment [here](https://github.com/vhirtham/weldx/pull/2))

**Comment by vhirtham** [comment]
I merged this one, so I can continue with the work on #497 . We can merge your other branch into master if we agreed on a solution. @CagtayFabry
